packages/protocol/src/utils/StreamPartitionID.ts

```diff
@@ -0,0 +1,154 @@
+    /** string key representing SPID */
+    public readonly key: string
+
+    protected readonly cachedSplit: [string, number] // cached split
+
+    /**
+     * @param id - stream id
+     * @param partition - stream partition
+     */
+    constructor(id: string, partition = 0) {
+        this.id = typeof id === 'string' ? id.toLowerCase() : id
+        this.partition = partition
+        this.validate()
+
+        // static values
+        const separator = (this.constructor as typeof SPID).SEPARATOR
+        this.key =`${this.id}${separator}${this.partition}`
+        this.cachedSplit = [this.id, this.partition]
+        // prevent all mutation
+        Object.freeze(this)
+        Object.freeze(this.cachedSplit)
+    }
+
+    /**
+     * Throws if data is invalid.
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new Error('SPID validation failed: id must be non-empty string')
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new Error('SPID validation failed: paritition must be an integer >= 0 ')
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDish): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = (this.constructor as typeof SPID).from(other)
+        } catch (_err) {
+            // not equal if not valid
+            return false
+        }
+
+        // check if matches both id + partition
+        return (
+            this.id === otherSpid.id
+            && this.partition === otherSpid.partition
+        )
+    }
+
+    /**
+     * Convert to SPID if possible.
+     */
+    static from(spidish: SPIDish): SPID {
+        // return spid if already spid
+        if (spidish instanceof SPID) {
+            return spidish
+        }
+
+        // convert from string
+        if (typeof spidish === 'string') {
+            return SPID.fromString(spidish)
+        } else if (spidish && typeof spidish === 'object') {
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamId = spidish.streamId || spidish.id
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamPartition = spidish.streamPartition || spidish.partition
+            return SPID.from(`${streamId}${this.SEPARATOR}${streamPartition}`)
+        } else {
+            // TODO: add more conversions?
+            throw new Error(`SPID validation failed, input is malformed: ${spidish}`)
+        }
+    }
+
+    /**
+     * Convert String to SPID if possible.
+     */
+    static fromString(spidString: string): SPID {
+        const [id, parititionStr] = spidString.split(this.SEPARATOR)
+        return new SPID(id, Number.parseFloat(parititionStr) || 0)
```

Comment:
I'm inclined to think it should be an error if the passed string does not conform to the pattern `stream-id|partition` or if the partition is non-numeric. In its current form, it seems like it will accept such a string but just default the partition to be zero?

Reply:
yes perhaps

Reply:
The way I see it:

1. We are instantiating a representation of `streamId` _and_ `partition` here. Not too much to ask imo that both be supplied or add a `|0` to the end.

1. Catching accidents
```ts
function performWhatever(userId: string, streamId: string): void {
   const spid = SPID.from(streamId)
   doAsSomething(userId, spid)
}

...

performWhatever('streamId', '0xFFFF...') // whoops
```

Reply:
I think the intention was so the SPID parsing in the implementation of this would works out of the box: `client.subscribe(streamId)` and `client.subscribe({ streamId })`

Reply:
having an string marker/prefix that allows us to easily detect a streamId vs some other random string isn't a bad idea though

Reply:
The lack of `partition` has a context-dependent interpretation, right? Given stream `s` that has 4 partitions what you'd really want is something as follows.

Subscribing:
```ts
SPID.from('s') // returns [new SPID('s', 0), new SPID('s', 1), new SPID('s', 2), new SPID('s', 3)]
```

Publishing
```ts
SPID.from('s') // returns new SPID('s', randomIntegerBetween(0,3))
```

I think interpreting it as zero automatically is not what is wanted in our use cases.

Reply:
done